### PR TITLE
Python: slight improvements in kernel and function result content, sample of local function calling with Nexus Raven

### LINKED ISSUE #6804: Python: Add support for kernel.InvokePromptAsync similar to C# counterpart.
Currently we have Kernel examples that are very simple using C#:

```csharp
Kernel kernel = Kernel.CreateBuilder()
            .AddOpenAIChatCompletion(modelId: “gpt-4o”, apiKey: “…”)
            .Build();

Console.WriteLine(await kernel.InvokePromptAsync("Why is the sky blue?"));
```

In python there's no way to use it as `function_name` and `plugin_name` are required arguments and need to be created manually with a `prompttemplateconfig`.

Desired but not supported today
```python
kernel = Kernel()
kernel.add_service(OpenAIChatCompletion(ai_model_id="gpt-4o", api_key="..."))

print(asyncio.run(kernel.invoke_prompt("Why is the sky blue")))
```
Currently to achieve a result all those lines are required in python:

```python
kernel = Kernel()
kernel.add_service(OpenAIChatCompletion(ai_model_id="gpt-4o", api_key="..."))
prompt_template_config = PromptTemplateConfig(template="why is the sky blue")
tldr_function = kernel.add_function(function_name="tldrFunction",plugin_name="tldrPlugin", prompt_template_config = prompt_template_config)
result = asyncio.run(kernel.invoke(tldr_function))
print(result)
```

### LINKED ISSUE #6944: Python: Create a sample showing how to perform function calling with a local model
Use Nexus Raven as the model for these samples

- [ ] Single function
- [ ] Parallel calls
- [ ] No calls
- [ ] Multiple functions
- [ ] Nested functions

### PR DESCRIPTION
**Pull request by eavanvalkenburg**
### Motivation and Context

This introduces a sample that uses function calling with a local model, Nexus Raven V2, in this case using the API, but the same model can be loaded using the Hugging Face Connection, the class here is highly adapted and specific to Nexus including parsing the returned function call value, into FunctionCallContent and then running that.
It then returns one or more ChatMessageContents with the function call content and function result content which can then be shared with a OpenAI call.

Error handling improvements:

* [`python/semantic_kernel/contents/chat_history.py`](diffhunk://#diff-b57a30f97fec1a91bbf6f7d15584473a8f0cf409797bdcd90d8c80bedc670892L266-R267): Improved error logging when parsing prompts as XML fails.
* [`python/semantic_kernel/contents/function_result_content.py`](diffhunk://#diff-3b69e6aedf4c29038cdff0ac99602b70583ab6c85f7bd777b08ad96b7fa48d87L7-R10): Removed field validator for `result` and allow result to be any value, instead of casting everything to string. [[1]](diffhunk://#diff-3b69e6aedf4c29038cdff0ac99602b70583ab6c85f7bd777b08ad96b7fa48d87L7-R10) [[2]](diffhunk://#diff-3b69e6aedf4c29038cdff0ac99602b70583ab6c85f7bd777b08ad96b7fa48d87L50-R51) [[3]](diffhunk://#diff-3b69e6aedf4c29038cdff0ac99602b70583ab6c85f7bd777b08ad96b7fa48d87L63-L69) [[4]](diffhunk://#diff-3b69e6aedf4c29038cdff0ac99602b70583ab6c85f7bd777b08ad96b7fa48d87R92-R115)

Minor changes:

* [`python/semantic_kernel/core_plugins/math_plugin.py`](diffhunk://#diff-632444474bd3ea991b4409dfd0ab194e0789c73617bef4a3dcae4db5c4b0b6c0L32-R38): Removed some docstrings for clarity. [[1]](diffhunk://#diff-632444474bd3ea991b4409dfd0ab194e0789c73617bef4a3dcae4db5c4b0b6c0L32-R38) [[2]](diffhunk://#diff-632444474bd3ea991b4409dfd0ab194e0789c73617bef4a3dcae4db5c4b0b6c0L55-R47)
* [`python/semantic_kernel/functions/kernel_function_from_prompt.py`](diffhunk://#diff-eabd53da2ec241cce6f92de85283065e7b478d62cefb52817c5b29ff07e28189L173-R174): Changed the order of condition checks for better readability.
* [`python/semantic_kernel/kernel.py`](diffhunk://#diff-3db6fe98818eda0b824788901c718d35eac8b13d0ff428d3dfd417866d922d3aR9-R10): Added functionality to update kernel arguments in `invoke_stream` method and made `function_name` and `plugin_name` optional in `invoke_prompt` method. [[1]](diffhunk://#diff-3db6fe98818eda0b824788901c718d35eac8b13d0ff428d3dfd417866d922d3aR9-R10) [[2]](diffhunk://#diff-3db6fe98818eda0b824788901c718d35eac8b13d0ff428d3dfd417866d922d3aR29) [[3]](diffhunk://#diff-3db6fe98818eda0b824788901c718d35eac8b13d0ff428d3dfd417866d922d3aR119-R120) [[4]](diffhunk://#diff-3db6fe98818eda0b824788901c718d35eac8b13d0ff428d3dfd417866d922d3aL195-R202) [[5]](diffhunk://#diff-3db6fe98818eda0b824788901c718d35eac8b13d0ff428d3dfd417866d922d3aL209-R216)

Closes #6804
Closes #6944

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/README.md">Py3.10 Test Coverage</a><details><summary>Python 3.10 Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>semantic_kernel</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/kernel.py">kernel.py</a></td><td>95</td><td>38</td><td>60%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/kernel.py#L127">127</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/kernel.py#L134-L143">134&ndash;143</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/kernel.py#L181-L189">181&ndash;189</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/kernel.py#L260-L296">260&ndash;296</a></td></tr><tr><td colspan="5"><b>semantic_kernel/contents</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py">function_result_content.py</a></td><td>62</td><td>13</td><td>79%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L56">56</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L61">61</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L67">67</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L72">72</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L87">87</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L99">99</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L113">113</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L118">118</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/contents/function_result_content.py#L125-L129">125&ndash;129</a></td></tr><tr><td colspan="5"><b>semantic_kernel/functions</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/functions/kernel_function_from_prompt.py">kernel_function_from_prompt.py</a></td><td>162</td><td>7</td><td>96%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/functions/kernel_function_from_prompt.py#L165-L166">165&ndash;166</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/functions/kernel_function_from_prompt.py#L187">187</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/functions/kernel_function_from_prompt.py#L207">207</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/functions/kernel_function_from_prompt.py#L232">232</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/functions/kernel_function_from_prompt.py#L252">252</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/7e021956599f776336878f870eaab622bdf8bc9b/python/semantic_kernel/functions/kernel_function_from_prompt.py#L335">335</a></td></tr><tr><td><b>TOTAL</b></td><td><b>6682</b></td><td><b>754</b></td><td><b>89%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python 3.10 Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 1561 | 1 :zzz: | 0 :x: | 0 :fire: | 24.111s :stopwatch: |

File: python/samples/concepts/auto_function_calling/nexus_raven.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        **kwargs: Any,
+    ) -> list["ChatMessageContent"]:
+        """Creates a chat message content with a function call content within.
+
+        Uses the text content and and parses it into a function call content."""
+        result = await self.get_text_contents(
+            prompt=chat_history.messages[-1].content,
+            settings=settings,
+        )
+        fcc, frc = await self._execute_function_calls(result[0], **kwargs)
```
Comment (moonbox3) [inline python/samples/concepts/auto_function_calling/nexus_raven.py:91]: As a sample, what do you think about making these names for explicit? Some users may not know what `fcc` or `frc` stand for.

File: python/semantic_kernel/kernel.py [hunk 2]
```diff
@@ -222,7 +227,7 @@ async def invoke_prompt(
             raise TemplateSyntaxError("The prompt is either null or empty.")
 
         function = KernelFunctionFromPrompt(
-            function_name=function_name,
+            function_name=function_name or generate_random_ascii_name(),
```
Comment (moonbox3) [inline python/semantic_kernel/kernel.py:230]: Thanks for handling this here. Could you add `Closes #6804` to your PR description, please?

**Review by moonbox3 — APPROVED** [review]

File: python/samples/concepts/auto_function_calling/nexus_raven.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+            by_alias=True,
+        )
+
+
+class NexusRavenCompletion(TextCompletionClientBase, ChatCompletionClientBase):
+    """To use this class, you should have installed the ``huggingface_hub`` package, and
+    the environment variable ``HUGGINGFACEHUB_API_TOKEN`` set with your API token,
+    or given as a named parameter to the constructor."""
+
+    client: AsyncInferenceClient
```
Comment (TaoChenOSU) [inline python/samples/concepts/auto_function_calling/nexus_raven.py:64]: Any chance this will become a hugging face connector?

Comment (eavanvalkenburg) [reply python/samples/concepts/auto_function_calling/nexus_raven.py:79]: Not in it's current form, I was thinking we should have a look to see the best approach for huggingface_hub models, whether as a subtype to the current one or as a seperate one, this one is highly specialized for this scenario, that is why I named it NexusRavenCompletion and not something else!

**Review by alliscode — APPROVED** [review]

**Review by moonbox3 — APPROVED** [review]

**Pull request merged**